File: bindings/wasm/audit_trail_wasm/examples/src/03_update_metadata.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * ## Actors
 *
 * - **Admin client**: Creates the trail and sets up the MetadataAdmin role.
 * - **Metadata admin client**: Holds the MetadataAdmin capability and updates the trail's mutable
 *   status field. Has no record-write permissions.
 *
 * Demonstrates how to:
 * 1. Create a trail with immutable and updatable metadata.
 * 2. Delegate metadata updates through a dedicated MetadataAdmin role.
 * 3. Change and clear the trail's updatable metadata.
 * 4. Verify that immutable metadata never changes.
 */

import { CapabilityIssueOptions, PermissionSet } from "@iota/audit-trail/node";
import { strict as assert } from "assert";
import { getFundedClient, TEST_GAS_BUDGET } from "./util";

export async function updateMetadata(): Promise<void> {
    console.log("=== Audit Trail: Update Metadata ===\n");

    // `adminClient` creates the trail and delegates metadata updates.
    // `metadataAdminClient` holds the MetadataAdmin capability and updates the status.
    const adminClient = await getFundedClient();
    const metadataAdminClient = await getFundedClient();

    const { output: createdTrail } = await adminClient
        .createTrail()
        .withTrailMetadata("Shipment Processing", "Tracks the lifecycle of a warehouse shipment")
        .withUpdatableMetadata("Status: Draft")
        .withInitialRecordString("Shipment created", "event:created")
        .finish()
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const trailId = createdTrail.id;

    // Delegate metadata updates to a MetadataAdmin role.
    const metadataAdminRole = adminClient.trail(trailId).access().forRole("MetadataAdmin");
    await metadataAdminRole
        .create(PermissionSet.metadataAdminPermissions())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);
    await metadataAdminRole
        .issueCapability(new CapabilityIssueOptions(metadataAdminClient.senderAddress()))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const trailBeforeUpdate = await adminClient.trail(trailId).get();
    console.log("Before update:");
    console.log("  immutable =", trailBeforeUpdate.immutableMetadata);
    console.log("  updatable =", trailBeforeUpdate.updatableMetadata, "\n");

    // MetadataAdmin updates the mutable metadata.
    await metadataAdminClient
        .trail(trailId)
        .updateMetadata("Status: In Review")
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(metadataAdminClient);

    const trailAfterUpdate = await adminClient.trail(trailId).get();
    console.log("After update:");
    console.log("  immutable =", trailAfterUpdate.immutableMetadata);
    console.log("  updatable =", trailAfterUpdate.updatableMetadata, "\n");

    assert.equal(trailAfterUpdate.immutableMetadata?.name, "Shipment Processing");
    assert.equal(trailAfterUpdate.updatableMetadata, "Status: In Review");

    // MetadataAdmin clears the mutable metadata.
    await metadataAdminClient
        .trail(trailId)
        .updateMetadata(undefined)
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(metadataAdminClient);

    const trailAfterClear = await adminClient.trail(trailId).get();
    console.log("After clear:");
    console.log("  immutable =", trailAfterClear.immutableMetadata);
    console.log("  updatable =", trailAfterClear.updatableMetadata);

    assert.equal(trailAfterClear.immutableMetadata?.name, "Shipment Processing");
    assert.equal(trailAfterClear.updatableMetadata, undefined);
}

File: examples/audit-trail/03_update_metadata.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! ## Actors
//!
//! - **Admin client**: Creates the trail and sets up the MetadataAdmin role.
//! - **Metadata admin client**: Holds the MetadataAdmin capability and updates the trail's mutable status field. Has no
//!   record-write permissions.

use anyhow::{Result, ensure};
use audit_trail::core::types::{CapabilityIssueOptions, Data, ImmutableMetadata, InitialRecord, PermissionSet};
use examples::get_funded_audit_trail_client;
use product_common::core_client::CoreClient;

/// Demonstrates how to:
/// 1. Create a trail with immutable and updatable metadata.
/// 2. Delegate metadata updates through a dedicated `MetadataAdmin` role.
/// 3. Change and clear the trail's updatable metadata.
/// 4. Verify that immutable metadata never changes.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Audit Trail: Update Metadata ===\n");

    // Use separate clients so metadata updates are clearly delegated away from the creator.
    let admin_client = get_funded_audit_trail_client().await?;
    let metadata_admin_client = get_funded_audit_trail_client().await?;

    let immutable_metadata = ImmutableMetadata::new(
        "Shipment Processing".to_string(),
        Some("Tracks the lifecycle of a warehouse shipment".to_string()),
    );

    let created_trail = admin_client
        .create_trail()
        .with_trail_metadata(immutable_metadata.clone())
        .with_updatable_metadata("Status: Draft")
        .with_initial_record(InitialRecord::new(
            Data::text("Shipment created"),
            Some("event:created".to_string()),
            None,
        ))
        .finish()
        .build_and_execute(&admin_client)
        .await?
        .output;

    let trail_id = created_trail.trail_id;
    let metadata_admin_role = "MetadataAdmin";

    // The Admin capability in `admin_client`'s wallet authorizes role definition and capability issuance.
    admin_client
        .trail(trail_id)
        .access()
        .for_role(metadata_admin_role)
        .create(PermissionSet::metadata_admin_permissions(), None)
        .build_and_execute(&admin_client)
        .await?;

    admin_client
        .trail(trail_id)
        .access()
        .for_role(metadata_admin_role)
        .issue_capability(CapabilityIssueOptions {
            issued_to: Some(metadata_admin_client.sender_address()),
            valid_from_ms: None,
            valid_until_ms: None,
        })
        .build_and_execute(&admin_client)
        .await?;

    let trail_before_update = admin_client.trail(trail_id).get().await?;
    println!(
        "Before update:\n  immutable = {:?}\n  updatable = {:?}\n",
        trail_before_update.immutable_metadata, trail_before_update.updatable_metadata
    );

    metadata_admin_client
        .trail(trail_id)
        .update_metadata(Some("Status: In Review".to_string()))
        .build_and_execute(&metadata_admin_client)
        .await?;

    let trail_after_update = admin_client.trail(trail_id).get().await?;
    println!(
        "After update:\n  immutable = {:?}\n  updatable = {:?}\n",
        trail_after_update.immutable_metadata, trail_after_update.updatable_metadata
    );

    ensure!(trail_after_update.immutable_metadata == Some(immutable_metadata.clone()));
    ensure!(trail_after_update.updatable_metadata.as_deref() == Some("Status: In Review"));

    metadata_admin_client
        .trail(trail_id)
        .update_metadata(None)
        .build_and_execute(&metadata_admin_client)
        .await?;

    let trail_after_clear = admin_client.trail(trail_id).get().await?;
    println!(
        "After clear:\n  immutable = {:?}\n  updatable = {:?}",
        trail_after_clear.immutable_metadata, trail_after_clear.updatable_metadata
    );

    ensure!(trail_after_clear.immutable_metadata == Some(immutable_metadata));
    ensure!(trail_after_clear.updatable_metadata.is_none());

    Ok(())
}
